File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章完全二叉树的权值.py
#完全二叉树的权值  2019年第十届省赛C/C++A组
# lanqiaoOJ 183

p = []
for i in range(0, 33):   #32层
    p.append(2 ** i)     #每层的个数：p[]=1,2,4,8,16,... 
n = int(input())
a = input().split(" ") 
sum = [0] * 32   #记录每层的和
for i in range(0, n):
    for j in range(0, len(p)):
        if i + 1 >= p[j] and i + 1 < p[j + 1]:  #计算每层的和
            sum[j] += int(a[i]) 
maxx, mindeep = -1, -1
for i in range(0, len(sum)):
    if maxx < sum[i]:
        maxx, mindeep = sum[i], i
print(mindeep + 1)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章完全二叉树的权值.cpp
//完全二叉树的权值  2019年第十届省赛C/C++A组
// lanqiaoOJ 183 

#include<bits/stdc++.h>
using namespace std;
int a[101000];
int p(int deep){                 //返回第deep层的数量，1,2,4,8,16,...
    return pow(2,deep);
}
int main(){
    int n,mindeep=1,i=2,deep=2;
    scanf("%d %d",&n,&a[1]);
    long long maxx = a[1];       //最大和，初值为第一层的和
    for(i=2;i<=n;i++) scanf("%d",&a[i]);
    i = 2;                       //第二层从a[2]开始
    while(i<=n){
        long long sum=0;
        for(;i<p(deep)&&i<=n;i++)  sum+=a[i];   //计算第deep层的和
        if(sum > maxx) maxx = sum, mindeep = deep;
        deep++;
  }
   cout << mindeep;
   return 0;
}
